File: src/game/graphics/skybox.rs
/*
 * Copyright (c) 2021 LambdAurora <aurora42lambda@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::path::Path;

use super::cube_map::CubeMapTexture;
use super::Shader;

const SKYBOX_VERTICES: [f32; 108] = [
    // Positions
    -1.0, 1.0, -1.0,
    -1.0, -1.0, -1.0,
    1.0, -1.0, -1.0,
    1.0, -1.0, -1.0,
    1.0, 1.0, -1.0,
    -1.0, 1.0, -1.0,
    -1.0, -1.0, 1.0,
    -1.0, -1.0, -1.0,
    -1.0, 1.0, -1.0,
    -1.0, 1.0, -1.0,
    -1.0, 1.0, 1.0,
    -1.0, -1.0, 1.0,
    1.0, -1.0, -1.0,
    1.0, -1.0, 1.0,
    1.0, 1.0, 1.0,
    1.0, 1.0, 1.0,
    1.0, 1.0, -1.0,
    1.0, -1.0, -1.0,
    -1.0, -1.0, 1.0,
    -1.0, 1.0, 1.0,
    1.0, 1.0, 1.0,
    1.0, 1.0, 1.0,
    1.0, -1.0, 1.0,
    -1.0, -1.0, 1.0,
    -1.0, 1.0, -1.0,
    1.0, 1.0, -1.0,
    1.0, 1.0, 1.0,
    1.0, 1.0, 1.0,
    -1.0, 1.0, 1.0,
    -1.0, 1.0, -1.0,
    -1.0, -1.0, -1.0,
    -1.0, -1.0, 1.0,
    1.0, -1.0, -1.0,
    1.0, -1.0, -1.0,
    -1.0, -1.0, 1.0,
    1.0, -1.0, 1.0
];

struct VAOHolder {
    pub vao: u32,
    pub vbo: u32,
}

impl VAOHolder {
    pub fn bind(&mut self) {
        unsafe {
            if SKYBOX_VAO.vao == 0 {
                gl::GenVertexArrays(1, &mut self.vao);
                gl::GenBuffers(1, &mut self.vbo);
                gl::BindVertexArray(self.vao);
                gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
                let size = (SKYBOX_VERTICES.len() * std::mem::size_of::<f32>()) as isize;
                let data = &SKYBOX_VERTICES[0] as *const f32 as *const std::ffi::c_void;
                gl::BufferData(gl::ARRAY_BUFFER, size, data, gl::STATIC_DRAW);
                gl::EnableVertexAttribArray(0);
                gl::VertexAttribPointer(0, 3, gl::FLOAT, gl::FALSE,
                                        (std::mem::size_of::<f32>() * 3) as i32, std::ptr::null());
            }

            gl::BindVertexArray(self.vao);
        }
    }
}

impl Drop for VAOHolder {
    fn drop(&mut self) {
        if self.vao != 0 {
            unsafe {
                gl::DeleteVertexArrays(1, &self.vao);
                gl::DeleteBuffers(1, &self.vbo);
            }
        }
    }
}

static mut SKYBOX_VAO: VAOHolder = VAOHolder { vao: 0, vbo: 0 };

/// Represents a skybox.
pub struct Skybox {
    pub cube_map: CubeMapTexture,
    pub shader: Shader,
}

impl Skybox {
    pub fn new(texture: CubeMapTexture, shader: Shader) -> Self {
        let mut skybox = Skybox { cube_map: texture, shader };
        skybox.shader.use_program();
        skybox.shader.set_i32("skybox", 0);
        skybox.scale(1.0);
        skybox
    }

    /// Returns a new skybox using the texture and loads the default skybox shader.
    pub fn load(texture: CubeMapTexture) -> Result<Self, String> {
        let shader = Shader::load(Path::new("assets/shaders/skybox.vsh"), Path::new("assets/shaders/skybox.fsh"))?;
        Ok(Skybox::new(texture, shader))
    }

    /// Sets the scale of the skybox.
    pub fn scale(&mut self, scale: f32) {
        self.shader.use_program();
        self.shader.set_float("scale", scale);
    }

    /// Draws the skybox.
    pub fn draw(&self) {
        unsafe { gl::DepthFunc(gl::LEQUAL); }
        self.shader.use_program();
        unsafe {
            SKYBOX_VAO.bind();
            gl::ActiveTexture(gl::TEXTURE0);
        }
        self.cube_map.bind();
        unsafe {
            gl::DrawArrays(gl::TRIANGLES, 0, 36);
            gl::BindVertexArray(0);
            gl::DepthFunc(gl::LESS);
        }
    }
}

File: src/main/java/dev/lambdaurora/res_errare/render/Skybox.java
/*
 * Copyright (c) 2021 LambdAurora <aurora42lambda@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package dev.lambdaurora.res_errare.render;

import dev.lambdaurora.res_errare.Constants;
import dev.lambdaurora.res_errare.render.array.VertexArray;
import dev.lambdaurora.res_errare.render.array.VertexLayout;
import dev.lambdaurora.res_errare.render.buffer.BufferTarget;
import dev.lambdaurora.res_errare.render.buffer.GraphicsBuffer;
import dev.lambdaurora.res_errare.render.shader.Shader;
import dev.lambdaurora.res_errare.render.shader.ShaderProgram;
import dev.lambdaurora.res_errare.render.shader.ShaderType;
import dev.lambdaurora.res_errare.render.texture.CubeMapTexture;
import dev.lambdaurora.res_errare.system.GL;
import dev.lambdaurora.res_errare.util.Identifier;
import dev.lambdaurora.res_errare.util.Result;

import java.util.List;

/**
 * Represents a skybox.
 */
public class Skybox implements AutoCloseable {
	public static final Identifier SHADER_ID = new Identifier(Constants.NAMESPACE, "skybox");
	private static final VertexLayout VAO_LAYOUT = new VertexLayout(List.of(VertexLayout.VEC3F_ELEMENT));
	private static final VertexArray SKYBOX_VAO = new VertexArray();
	private static GraphicsBuffer skyboxVbo;

	private final CubeMapTexture texture;
	private final ShaderProgram shader;

	public Skybox(CubeMapTexture texture, ShaderProgram shader) {
		this.texture = texture;
		this.shader = shader;

		this.shader.use();
		this.shader.setInt("skybox", 0);
		this.scale(1.f);
	}

	public static Result<Skybox, ShaderProgram.LinkageError> of(CubeMapTexture texture) {
		return ShaderProgram.builder()
				.shader(Shader.compile(ShaderType.FRAGMENT, SHADER_ID))
				.shader(Shader.compile(ShaderType.VERTEX, SHADER_ID))
				.withCleanup()
				.build()
				.then(shader -> Result.ok(new Skybox(texture, shader)));
	}

	/**
	 * Scales the skybox.
	 *
	 * @param scale the new scale of the skybox
	 */
	public void scale(float scale) {
		this.shader.use();
		this.shader.setFloat("scale", scale);
	}

	/**
	 * Draws the skybox.
	 */
	public void draw() {
		GL.get().depthFunc(GL.GL11.LEQUAL);
		this.shader.use();
		GL.get().activeTexture(GL.GL13.TEXTURE0);
		this.texture.bind();
		bindSkyboxVao();
		SKYBOX_VAO.draw(GeometricPrimitive.TRIANGLE_STRIP, 0, VAO_LAYOUT, VERTICES.length);
		SKYBOX_VAO.finish();
		this.texture.unbind();
		GL.get().depthFunc(GL.GL11.LESS);
	}

	@Override
	public void close() {
		this.shader.close();
		this.texture.close();
	}

	private static void bindSkyboxVao() {
		boolean valid = SKYBOX_VAO.isValid();

		SKYBOX_VAO.bind();

		if (!valid) {
			skyboxVbo = GraphicsBuffer.ofStatic(BufferTarget.ARRAY, VERTICES);
			SKYBOX_VAO.useLayout(VAO_LAYOUT);
			skyboxVbo.unbind();
		}
	}

	public static void terminate() {
		SKYBOX_VAO.close();
		if (skyboxVbo != null)
			skyboxVbo.close();
		skyboxVbo = null;
	}

	private static final float[] VERTICES = new float[]{
			// Positions
			// 1. Quad (complete)
			-1.f, -1.f, 1.f,
			-1.f, -1.f, -1.f,
			-1.f, 1.f, 1.f,
			-1.f, 1.f, -1.f,
			// 2. Quad
			1.f, 1.f, 1.f,
			1.f, 1.f, -1.f,
			// 3. Quad
			1.f, -1.f, 1.f,
			1.f, -1.f, -1.f,
			// | Loop 1 to stop rendering
			1.f, -1.f, -1.f,
			1.f, 1.f, -1.f,
			// | Loop 2 to reposition new start point
			// | 4. Quad (complete)
			1.f, 1.f, -1.f,
			-1.f, 1.f, -1.f,
			1.f, -1.f, -1.f,
			-1.f, -1.f, -1.f,
			// 5. Quad
			1.f, -1.f, 1.f,
			-1.f, -1.f, 1.f,
			// 6. Quad
			1.f, 1.f, 1.f,
			-1.f, 1.f, 1.f,
	};
}
